### DISCUSSION TITLE
useQuery with selector is causing infinite requests / re renders

### DISCUSSION THREAD

**Discussion by rynebenson** [TanStack/query#3308] [Q&A]
I'm working on a NextJS project and I have a large list of items that can be filtered, updated, and deleted. I'm fetching the items at the page level like this

```
export function useLibraryItems() {
  return useQuery(['items'], () => API.FETCH_LIBRARY_ITEMS().then(response => response.data))
}

// in my component
const items = useLibraryItems()
```
I'm rendering the items in a virtual list which looks like this

```
const ItemRow = ({ index, style }) => (
    <Item
      key={index}
      id={items.data[index]._id}
      style={style}
      hasBottomBorder={index + 1 === items.data.length}
    />
 )

{items.status === 'loading' ? <div className={styles.spinner} /> : null}

{items.status === 'success' ? (
  items.data.length > 0 ? (
    <div className={styles.item_table}>
      <ItemHeading
        marginRight={84}
     />

     <FixedSizeList
        height={virtualListDimensions.height}
        width={virtualListDimensions.width}
        itemSize={30}
        itemCount={items.data.length}
     >
        {ItemRow}
      </FixedSizeList>
    </div>
  ) : (
    <p className={styles.empty_items_message}>There are no global items</p>
  )
) : null}
```

My item selector looks like this

```
export const useLibraryItem = (id) => {
  return useQuery(['items'], () => API.FETCH_LIBRARY_ITEMS().then(response => response.data), { select: (items) => items.find(item => item._id === id) })
}
```
I'm calling this in my component like this

```
const item = useLibraryItem(id)

useEffect(() => {
    console.log(item)
}, [])
```
Here is the output of this code

<img width="484" alt="Screen Shot 2022-02-17 at 8 50 27 PM" src="https://user-images.githubusercontent.com/26049860/154602257-4ad4ca1d-4aa3-4762-8142-9ae8612a6633.png">

<img width="487" alt="Screen Shot 2022-02-17 at 8 50 05 PM" src="https://user-images.githubusercontent.com/26049860/154602271-f7dfd014-b809-4f86-9704-6864d72be144.png">

My goal of setting these selectors up is to achieve the ability to mutate the items individually while optimistically updating the items list. I have a mutation query that looks like this 

```
export function useUpdateLibraryItem() {
  let client = useQueryClient()

  return useMutation(args => API.UPDATE_LIBRARY_ITEM(args.id, args.params).then(response => response.data), {
    onMutate: async (args) => {
      await client.cancelQueries('items')

      let prev = client.getQueriesData('items')

      client.setQueriesData('items', items => [
        ...items.map(item => {
          if(item._id === args.id)
            return { ...item, ...args.params }
          else 
            return item
        })
      ])

      return { prev }
    }
  })
}
```

In my component I have inputs to update the item values and a button to save the updates. If I update 3 items and save the changes on one then the other items will reset to their original value. I believe if I can get the selectors working correctly it would run the mutation and only the current item that i ran the mutation on would update. Is there a bug here with react query selectors or am I missing something?

**Comment by TkDodo** [comment]
> I believe if I can get the selectors working correctly it would run the mutation and only the current item that i ran the mutation on would update. 

that is correct. can you reproduce the behaviour in a codesandbox please?

**Reply by rynebenson** [reply]
Sure, here's a simplified example of my app in a codesandbox https://codesandbox.io/s/empty-frog-q0w646?file=/src/components/Item.js . If you uncomment out query and console log in the Item.js file you'll see the behavior that I am describing

**Reply by rynebenson** [reply]
I just realized that the react-query selector is working correctly and that the virtual list is what's causing the app to re render infinitely.

**Reply by TkDodo** [reply]
no idea how `FixedSizeList` from `react-window`works, I haven't had problems with getting react-query to work well together with [react-virtual](https://react-virtual.tanstack.com/)

I also don't see any `select` in your example 🤷

**Reply by rynebenson** [reply]
The select is at the end of the libs/product.js file. I switched over to `react-virtual` and everything is working as expected which is amazing and I really appreciate you guys creating these awesome libraries! I'm not sure if I should mark your response as answer or close this discussions somehow, but I can do that. While the discussion is open though, I do have another question  that sprung from the selectors that you may be able to help me with. 

When I create an item I am adding it to the items list by an optimistic update which looks like this

```
export function useCreateLibraryItem() {
  let client = useQueryClient()

  return useMutation(params => API.CREATE_LIBRARY_ITEM(params).then(response => response.data), {
    onMutate: async (params) => {
      await client.cancelQueries('items')

      let prev = client.getQueryData('items')
  
      client.setQueryData('items', items => [params, ...items])
  
      return { prev }
    },
    onError: (error, newItem, context) => {
      client.setQueryData('items', context.prev)
    },
    onSettled: () => {
      client.invalidateQueries('items')
    }
  })
}
```
When this runs it is adding the new item to the list successfully. When it does that a new component is rendered onto the screen, which calls the 'items' query. Because it does this faster than the backend updates, the item is removed from the screen and then shows back up again after. Heres the query and item component 

```
export const useLibraryItem = (id) => {
  return useQuery(['items'], () => API.FETCH_LIBRARY_ITEMS().then(response => response.data), { select: (items) => items.find(item => item._id === id) })
}

export default function Item({ id }) {
  const item = useLibraryItem(id) // There is a new instance of Item from the optimistic update and this query recalls the 'items' query 

  return (
    <div>
      ...
    </div>
  )
}
```

If I log the items.data.length on the root component I notice this behavior which confirms whats happening

```
const items = useLibraryItems()

console.log(items.data?.length)

/*
7
8 -> optimistic item makes it eight
7 -> useLibraryItem() in `Item` gets called and 'items' query returns the list which doesn't have the new item saved yet
8 -> onSuccess is triggered and updates the 'items' query to show the list which has the new item saved
*/
```

**Reply by TkDodo** [reply]
yeah that's an interesting case. I guess it's independent of selectors, but it manifests with them because you use the list query again when a new item to the list is added. cancelQueries only makes sure that currently running queries do not interfere with the optimistic update, but that doesn't work for fetches that are triggered _after_ that obviously.

Two possible solutions come to mind:

1) You set a small `staleTime` on your query. Then, when the new query mounts, it sees the optimistically updated data as `fresh` and doesn't trigger a refetch
2) You disable the newly mounted query for as long as it has data that is marked as optimistic. I haven't thought this through and don't know if it really works, because once disabled, you need a point where you enable it again. That seems only doable with global client state that you write in `onMutate` and revert in `onSettled` of the mutation before you invalidate. Doesn't seem great, `staleTime` looks a lot better.

**Comment by AClimber** [comment]
I had the same problem just wrap `select` callback function with `React.useCallback`


```
export const useLibraryItem = (id) => {
  return useQuery(['items'], () => API.FETCH_LIBRARY_ITEMS().then(response => response.data), { select: React.useCallback((items) => items.find(item => item._id === id) }), []);
}
```

https://tkdodo.eu/blog/react-query-data-transformations#3-using-the-select-option

**Comment by louis-young** [comment]
Also, it looks like you're defining a component inside of another component. This will break any possible memoisation as it'll recreate the component every time the parent re-renders.
